File: test/jvm/com/cattailsw/nanidroid/NarArchiveCharacterizationTest.java
package com.cattailsw.nanidroid;

import static org.junit.Assert.assertArrayEquals;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import com.cattailsw.nanidroid.util.NarUtil;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.nio.charset.Charset;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;

/**
 * Characterizes only successful, trusted forced-id extraction of a bounded,
 * collision-free archive. Rejection and containment policy belong to D9b.
 */
public final class NarArchiveCharacterizationTest {
    private static final Charset ASCII = Charset.forName("US-ASCII");
    private static final byte[] INSTALL_DESCRIPTOR = ascii(
            "type,ghost\n"
                    + "directory,descriptor-ghost\n");
    private static final byte[] GHOST_DESCRIPTOR = ascii(
            "name,Seed Ghost\n"
                    + "sakura.name,Seed Sakura\n");
    private static final byte[] SHELL_DESCRIPTOR = ascii("name,Master Shell\n");
    private static final byte[] README = ascii("Synthetic forced-id archive.\n");
    private static final byte[] BINARY_PAYLOAD = new byte[] {
            (byte) 0x00,
            (byte) 0x7f,
            (byte) 0x80,
            (byte) 0xff,
    };

    @Rule
    public final TemporaryFolder temporaryFolder = new TemporaryFolder();

    @Test
    public void requiredMigrationInvariant_trustedForcedIdPreservesExactTreeAndBytes()
            throws Exception {
        File sourceRoot = temporaryFolder.newFolder("source");
        File installRoot = temporaryFolder.newFolder("install");
        File archive = new File(sourceRoot, "synthetic.nar");
        writeArchive(archive);

        boolean installed = NarUtil.readNarArchive(
                archive.getAbsolutePath(),
                installRoot.getAbsolutePath(),
                "seed-ghost");

        assertTrue(installed);
        assertArrayEquals(new String[] {"seed-ghost"}, sortedNames(installRoot));
        assertFalse(new File(installRoot, "descriptor-ghost").exists());
        assertEquals(
                Arrays.asList(
                        "seed-ghost/",
                        "seed-ghost/ghost/",
                        "seed-ghost/ghost/master/",
                        "seed-ghost/ghost/master/data/",
                        "seed-ghost/ghost/master/data/payload.bin",
                        "seed-ghost/ghost/master/descript.txt",
                        "seed-ghost/install.txt",
                        "seed-ghost/readme.txt",
                        "seed-ghost/shell/",
                        "seed-ghost/shell/master/",
                        "seed-ghost/shell/master/descript.txt"),
                sortedTree(installRoot));

        File installedRoot = new File(installRoot, "seed-ghost");
        assertArrayEquals(
                INSTALL_DESCRIPTOR,
                readBytes(new File(installedRoot, "install.txt")));
        assertArrayEquals(
                GHOST_DESCRIPTOR,
                readBytes(new File(installedRoot, "ghost/master/descript.txt")));
        assertArrayEquals(
                SHELL_DESCRIPTOR,
                readBytes(new File(installedRoot, "shell/master/descript.txt")));
        assertArrayEquals(
                README,
                readBytes(new File(installedRoot, "readme.txt")));
        byte[] installedPayload = readBytes(
                new File(installedRoot, "ghost/master/data/payload.bin"));
        assertArrayEquals(BINARY_PAYLOAD, installedPayload);
        assertEquals(
                "89273d2f70b93285bb7ddb4bcee86a5347ca7159352e3cbdd20c23e9d1e507d3",
                sha256(installedPayload));
    }

    private static void writeArchive(File archive) throws Exception {
        ZipOutputStream output = new ZipOutputStream(new FileOutputStream(archive));
        try {
            // Deliberately avoid descriptor-first ordering.
            writeEntry(output, "ghost/master/data/payload.bin", BINARY_PAYLOAD);
            writeEntry(output, "readme.txt", README);
            writeEntry(output, "ghost/master/descript.txt", GHOST_DESCRIPTOR);
            writeEntry(output, "install.txt", INSTALL_DESCRIPTOR);
            writeEntry(output, "shell/master/descript.txt", SHELL_DESCRIPTOR);
        } finally {
            output.close();
        }
    }

    private static void writeEntry(
            ZipOutputStream output,
            String name,
            byte[] content) throws Exception {
        output.putNextEntry(new ZipEntry(name));
        output.write(content);
        output.closeEntry();
    }

    private static String[] sortedNames(File directory) {
        String[] names = directory.list();
        if (names == null) {
            return new String[0];
        }
        Arrays.sort(names);
        return names;
    }

    private static List<String> sortedTree(File root) {
        List<String> paths = new ArrayList<String>();
        collectTree(root, "", paths);
        Collections.sort(paths);
        return paths;
    }

    private static void collectTree(
            File directory,
            String prefix,
            List<String> paths) {
        File[] children = directory.listFiles();
        if (children == null) {
            return;
        }
        Arrays.sort(children, new Comparator<File>() {
            @Override
            public int compare(File left, File right) {
                return left.getName().compareTo(right.getName());
            }
        });
        for (File child : children) {
            String relative = prefix + child.getName();
            if (child.isDirectory()) {
                paths.add(relative + "/");
                collectTree(child, relative + "/", paths);
            } else {
                paths.add(relative);
            }
        }
    }

    private static byte[] readBytes(File file) throws Exception {
        FileInputStream input = new FileInputStream(file);
        try {
            ByteArrayOutputStream output = new ByteArrayOutputStream();
            byte[] buffer = new byte[256];
            int count;
            while ((count = input.read(buffer)) != -1) {
                output.write(buffer, 0, count);
            }
            return output.toByteArray();
        } finally {
            input.close();
        }
    }

    private static String sha256(byte[] content) throws Exception {
        byte[] digest = MessageDigest.getInstance("SHA-256").digest(content);
        StringBuilder result = new StringBuilder(digest.length * 2);
        for (byte value : digest) {
            result.append(String.format("%02x", value & 0xff));
        }
        return result.toString();
    }

    private static byte[] ascii(String value) {
        return value.getBytes(ASCII);
    }
}

File: src/test/java/com/cattailsw/nanidroid/NarArchiveCharacterizationTest.kt
package com.cattailsw.nanidroid

import android.content.Intent
import com.cattailsw.nanidroid.util.NarUtil
import com.cattailsw.nanidroid.install.NarLocalArchiveStager
import org.junit.Assert
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.nio.charset.Charset
import java.security.MessageDigest
import java.util.Arrays
import java.util.Collections
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream

/**
 * Characterizes only successful, trusted forced-id extraction of a bounded,
 * collision-free archive. Rejection and containment policy belong to D9b.
 */
class NarArchiveCharacterizationTest {
    @Rule
    @JvmField
    val temporaryFolder: TemporaryFolder = TemporaryFolder()

    @Test
    fun externalArchiveIntent_acceptsOnlyGrantedContentArchives() {
        Assert.assertTrue(
            ArchiveIntentAdapter.accepts(
                Intent.ACTION_VIEW,
                "content",
                "application/x-nar",
                Intent.FLAG_GRANT_READ_URI_PERMISSION,
            ),
        )
        Assert.assertFalse(
            ArchiveIntentAdapter.accepts(
                Intent.ACTION_VIEW,
                "https",
                "application/x-nar",
                Intent.FLAG_GRANT_READ_URI_PERMISSION,
            ),
        )
        Assert.assertFalse(
            ArchiveIntentAdapter.accepts(
                Intent.ACTION_VIEW,
                "content",
                "application/x-nar",
                0,
            ),
        )
    }

    @Test
    fun temporaryArchiveStageFailureLeavesNoPrivateFile() {
        val directory = temporaryFolder.newFolder("temporary-archive")

        val result = NarLocalArchiveStager.stage(directory) { null }

        Assert.assertTrue(result is NarLocalArchiveStager.Result.Failed)
        Assert.assertTrue(directory.listFiles().isNullOrEmpty())
    }

    @Test
    @Throws(Exception::class)
    fun requiredMigrationInvariant_trustedForcedIdPreservesExactTreeAndBytes() {
        val sourceRoot = temporaryFolder.newFolder("source")
        val installRoot = temporaryFolder.newFolder("install")
        val archive = File(sourceRoot, "synthetic.nar")
        writeArchive(archive)

        val installed: Boolean = NarUtil.readNarArchive(
            archive.getAbsolutePath(),
            installRoot.getAbsolutePath(),
            "seed-ghost"
        )

        Assert.assertTrue(installed)
        Assert.assertArrayEquals(arrayOf<String>("seed-ghost"), sortedNames(installRoot))
        Assert.assertFalse(File(installRoot, "descriptor-ghost").exists())
        Assert.assertEquals(
            mutableListOf<String>(
                "seed-ghost/",
                "seed-ghost/ghost/",
                "seed-ghost/ghost/master/",
                "seed-ghost/ghost/master/data/",
                "seed-ghost/ghost/master/data/payload.bin",
                "seed-ghost/ghost/master/descript.txt",
                "seed-ghost/install.txt",
                "seed-ghost/readme.txt",
                "seed-ghost/shell/",
                "seed-ghost/shell/master/",
                "seed-ghost/shell/master/descript.txt"
            ),
            sortedTree(installRoot)
        )

        val installedRoot = File(installRoot, "seed-ghost")
        Assert.assertArrayEquals(
            INSTALL_DESCRIPTOR,
            readBytes(File(installedRoot, "install.txt"))
        )
        Assert.assertArrayEquals(
            GHOST_DESCRIPTOR,
            readBytes(File(installedRoot, "ghost/master/descript.txt"))
        )
        Assert.assertArrayEquals(
            SHELL_DESCRIPTOR,
            readBytes(File(installedRoot, "shell/master/descript.txt"))
        )
        Assert.assertArrayEquals(
            README,
            readBytes(File(installedRoot, "readme.txt"))
        )
        val installedPayload: ByteArray = readBytes(
            File(installedRoot, "ghost/master/data/payload.bin")
        )
        Assert.assertArrayEquals(BINARY_PAYLOAD, installedPayload)
        Assert.assertEquals(
            "89273d2f70b93285bb7ddb4bcee86a5347ca7159352e3cbdd20c23e9d1e507d3",
            sha256(installedPayload)
        )
    }

    companion object {
        private val ASCII: Charset = Charset.forName("US-ASCII")
        private val INSTALL_DESCRIPTOR: ByteArray = ascii(
            "type,ghost\n"
                    + "directory,descriptor-ghost\n"
        )
        private val GHOST_DESCRIPTOR: ByteArray = ascii(
            "name,Seed Ghost\n"
                    + "sakura.name,Seed Sakura\n"
        )
        private val SHELL_DESCRIPTOR: ByteArray = ascii("name,Master Shell\n")
        private val README: ByteArray = ascii("Synthetic forced-id archive.\n")
        private val BINARY_PAYLOAD = byteArrayOf(
            0x00.toByte(),
            0x7f.toByte(),
            0x80.toByte(),
            0xff.toByte(),
        )

        @Throws(Exception::class)
        private fun writeArchive(archive: File) {
            val output = ZipOutputStream(FileOutputStream(archive))
            try {
                // Deliberately avoid descriptor-first ordering.
                writeEntry(output, "ghost/master/data/payload.bin", BINARY_PAYLOAD)
                writeEntry(output, "readme.txt", README)
                writeEntry(output, "ghost/master/descript.txt", GHOST_DESCRIPTOR)
                writeEntry(output, "install.txt", INSTALL_DESCRIPTOR)
                writeEntry(output, "shell/master/descript.txt", SHELL_DESCRIPTOR)
            } finally {
                output.close()
            }
        }

        @Throws(Exception::class)
        private fun writeEntry(
            output: ZipOutputStream,
            name: String,
            content: ByteArray
        ) {
            output.putNextEntry(ZipEntry(name))
            output.write(content)
            output.closeEntry()
        }

        private fun sortedNames(directory: File): Array<String> {
            val names = directory.list()
            if (names == null) {
                return emptyArray()
            }
            Arrays.sort(names)
            return names.filterNotNull().toTypedArray()
        }

        private fun sortedTree(root: File): MutableList<String> {
            val paths: MutableList<String> = ArrayList<String>()
            collectTree(root, "", paths)
            Collections.sort<String>(paths)
            return paths
        }

        private fun collectTree(
            directory: File,
            prefix: String,
            paths: MutableList<String>
        ) {
            val children = directory.listFiles()
            if (children == null) {
                return
            }
            Arrays.sort<File>(children, object : Comparator<File> {
                override fun compare(left: File, right: File): Int {
                    return left.getName().compareTo(right.getName())
                }
            })
            for (child in children) {
                val relative = prefix + child.getName()
                if (child.isDirectory()) {
                    paths.add(relative + "/")
                    collectTree(child, relative + "/", paths)
                } else {
                    paths.add(relative)
                }
            }
        }

        @Throws(Exception::class)
        private fun readBytes(file: File): ByteArray {
            val input = FileInputStream(file)
            try {
                val output = ByteArrayOutputStream()
                val buffer = ByteArray(256)
                var count: Int
                while ((input.read(buffer).also { count = it }) != -1) {
                    output.write(buffer, 0, count)
                }
                return output.toByteArray()
            } finally {
                input.close()
            }
        }

        @Throws(Exception::class)
        private fun sha256(content: ByteArray): String {
            val digest = MessageDigest.getInstance("SHA-256").digest(content)
            val result = StringBuilder(digest.size * 2)
            for (value in digest) {
                result.append(String.format("%02x", value.toInt() and 0xff))
            }
            return result.toString()
        }

        private fun ascii(value: String): ByteArray {
            return value.toByteArray(ASCII)
        }
    }
}
